## Introduction

In this project we are using santander data, which is realted to various transactions of users to predict if they would make any transaction or not.

We are using naive bayes and gradient boost algorithms to predict the output and comparing their accuracies to find the best algorithm.

Naive Bayes- as the name suggests, it uses bayes theorem to classify the output and is widely used algorithm, in many real time problems such as to classify a disease based on previous history, classify if an email is spam or not etc.It uses the bayes theorem to classify the output based on the probability of the condition but it compares each feature individually without taking any inference from the previous feature.

In [1]:
#importing all necessary libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn import metrics
from sklearn.metrics import confusion_matrix 
from sklearn.metrics import accuracy_score 
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split


Our data consists of 200 different variables and one target variable, based on the 200 variables or features we are predicting if the output would be one or zero i.e if the customer would pay or not.

In [2]:
# import the dataset
df=pd.read_csv("Bank_train.csv")
df1=pd.read_csv("Bank_test.csv")


In [3]:
#to view the data
df.head()

ID_code  target    var_0   var_1    var_2   var_3    var_4   var_5   var_6  \
0  train_0       0   8.9255 -6.7863  11.9081  5.0930  11.4607 -9.2834  5.1187   
1  train_1       0  11.5006 -4.1473  13.8588  5.3890  12.3622  7.0433  5.6208   
2  train_2       0   8.6093 -2.7457  12.0805  7.8928  10.5825 -9.0837  6.9427   
3  train_3       0  11.0604 -2.1518   8.9522  7.1957  12.5846 -1.8361  5.8428   
4  train_4       0   9.8369 -1.4834  12.8746  6.6375  12.2772  2.4486  5.9405   

     var_7  ...  var_190  var_191  var_192  var_193  var_194  var_195  \
0  18.6266  ...   4.4354   3.9642   3.1364   1.6910  18.5227  -2.3978   
1  16.5338  ...   7.6421   7.7214   2.5837  10.9516  15.4305   2.0339   
2  14.6155  ...   2.9057   9.7905   1.6704   1.6858  21.6042   3.1417   
3  14.9250  ...   4.4666   4.7433   0.7178   1.4214  23.0347  -1.2706   
4  19.2514  ...  -1.4905   9.5214  -0.1508   9.1942  13.2876  -1.5121   

   var_196  var_197  var_198  var_199  
0   7.8784   8.5635  12.7803  -1.0914  
1   8.1267   8.7889  18.3560   1.9518  
2  -6.5213   8.2675  14.7222   0.3965  
3  -2.9275  10.2922  17.9697  -8.9996  
4   3.9267   9.5031  17.9974  -8.8104  

[5 rows x 202 columns]

In [21]:
#Splitting the data into train and validation set
X1, Xvalid1, y1, yvalid1 = train_test_split(df, df.target, test_size=0.2)
print(X1.shape, y1.shape)
print(Xvalid1.shape, yvalid1.shape)

(160000, 202) (160000,)
(40000, 202) (40000,)


In [24]:
#here we are removing the target variable and taking the rest of them into x
X = X1.iloc[:,2:]
X_test=df1.iloc[:,1:]

In [28]:
Xvalid=Xvalid1.iloc[:,2:]
Xvalid.head()

var_0   var_1    var_2    var_3    var_4    var_5   var_6    var_7  \
131874   7.0240 -0.4980   8.9893   6.3568  10.1176   3.4995  3.7518  13.6428   
99433   10.0977  0.0324  12.3161   4.6699  10.5866 -12.7604  5.6334  17.4549   
106335   6.7115 -1.4901   7.2993  10.3569  11.1072 -15.0697  5.0655  18.6770   
41345   12.8994 -0.5779   9.0916   5.4894  11.9639 -19.2964  4.3991  17.8693   
154358  12.6760  3.3565  16.0442   7.2954   9.8858  -2.3777  5.8290  11.4296   

         var_8   var_9  ...  var_190  var_191  var_192  var_193  var_194  \
131874  0.1663  8.3179  ...  -3.7363  14.9481   0.9657  10.0847  17.1863   
99433   0.2002  8.6214  ...  10.3313   9.5215   3.1408   5.4062  19.4232   
106335  0.7786  9.1606  ...  -4.3652   8.8620   3.9670   0.7728  22.5873   
41345   4.8756  8.5959  ...   3.5126   3.1761  -0.5557   5.0394  11.3876   
154358 -2.9712  8.1556  ...   3.2732   8.0776   5.0450   9.5620  19.1064   

        var_195  var_196  var_197  var_198  var_199  
131874   0.6374  -0.5249   8.0139  14.6043   2.9617  
99433    1.5651  -4.9051   8.1672  17.0002   0.0602  
106335   2.0498  -2.1128  10.6504  16.7789   9.1361  
41345   -0.4160   1.7931   9.9067  15.8607 -11.4486  
154358  -2.0474   5.1863   9.3115  16.9803 -20.1300  

[5 rows x 200 columns]

In [25]:
#checking the shape of X
X.shape

(160000, 200)

In [29]:
#checking the shape of validation set
print(Xvalid.shape)

(40000, 200)


In [11]:
#view of our data in validation set
Xvalid.head()

var_0   var_1    var_2   var_3    var_4   var_5   var_6    var_7   var_8  \
0   8.9255 -6.7863  11.9081  5.0930  11.4607 -9.2834  5.1187  18.6266 -4.9200   
1  11.5006 -4.1473  13.8588  5.3890  12.3622  7.0433  5.6208  16.5338  3.1468   
2   8.6093 -2.7457  12.0805  7.8928  10.5825 -9.0837  6.9427  14.6155 -4.9193   
3  11.0604 -2.1518   8.9522  7.1957  12.5846 -1.8361  5.8428  14.9250 -5.8609   
4   9.8369 -1.4834  12.8746  6.6375  12.2772  2.4486  5.9405  19.2514  6.2654   

    var_9  ...  var_190  var_191  var_192  var_193  var_194  var_195  var_196  \
0  5.7470  ...   4.4354   3.9642   3.1364   1.6910  18.5227  -2.3978   7.8784   
1  8.0851  ...   7.6421   7.7214   2.5837  10.9516  15.4305   2.0339   8.1267   
2  5.9525  ...   2.9057   9.7905   1.6704   1.6858  21.6042   3.1417  -6.5213   
3  8.2450  ...   4.4666   4.7433   0.7178   1.4214  23.0347  -1.2706  -2.9275   
4  7.6784  ...  -1.4905   9.5214  -0.1508   9.1942  13.2876  -1.5121   3.9267   

   var_197  var_198  var_199  
0   8.5635  12.7803  -1.0914  
1   8.7889  18.3560   1.9518  
2   8.2675  14.7222   0.3965  
3  10.2922  17.9697  -8.9996  
4   9.5031  17.9974  -8.8104  

[5 rows x 200 columns]

In [30]:
#view of our data in train
X.head()

var_0   var_1    var_2    var_3    var_4    var_5   var_6    var_7  \
95262    9.9788 -8.3247   3.0668   4.1455   7.7924 -11.0567  6.4396  18.0775   
152870  10.9336  2.8859  15.3637  10.0805  12.5866 -10.4935  5.5890  15.0074   
133191   7.5925 -4.2309  12.4876   4.3598  12.6592 -14.8118  5.1159  17.8710   
51534   19.0715 -2.5063   8.8157   8.0177  12.0444   1.0364  4.6243  13.4528   
155413  14.6511 -6.4685  11.3598   5.6296  12.2257 -12.6634  6.8864  12.7346   

         var_8   var_9  ...  var_190  var_191  var_192  var_193  var_194  \
95262  -0.0275  9.0580  ...   6.9477   6.9868   0.7769  12.2067  19.1382   
152870 -1.9336  8.4996  ...   4.6502  10.5889   1.8718  -5.9628  22.4242   
133191  0.8528  7.6838  ...   1.2323   4.8060   1.1740  -2.3315  19.8882   
51534   5.2100  7.7623  ...   3.5853   6.8281  -0.2866   2.5941  12.8357   
155413  3.6726  7.5428  ...  11.9915   4.9920   0.3943   9.8817  16.4500   

        var_195  var_196  var_197  var_198  var_199  
95262   -2.0168  -6.0460   8.7234  15.8390 -18.6267  
152870   0.8031  -1.1052   7.7015  19.1582  -1.6164  
133191   0.4979   1.2373   9.9944  17.0772 -16.2193  
51534   -1.1218  -7.1058   7.9326  12.1268  -5.2001  
155413  -2.6252   5.5718   8.6786   9.6359   2.1737  

[5 rows x 200 columns]

### Naive Bayes
Here we are using the naive bayes model to evalute and predict our outcomes.

In [31]:
#Import Gaussian Naive Bayes model
from sklearn.naive_bayes import GaussianNB

#Create a Gaussian Classifier
nb = GaussianNB()
#Train the model using the training sets
a=nb.fit(X, y)
#Predict the response for test dataset
y_pred = nb.predict(X_test)
y_valid_pred= nb.predict(Xvalid)

In [32]:
#validation accuracy for Naive Bayes
print("Accuracy:",metrics.accuracy_score(yvalid1,y_valid_pred))

Accuracy: 0.9003


In [33]:
#confusion matrix for Naive Bayes 
c_m1 =confusion_matrix(yvalid1,y_valid_pred) 
c_m1

array([[36012,     0],
       [ 3988,     0]])

### Gradient Boost
Below we are using Gradient boost algorithm to evaluate our model and predict the desired output.

In Gradient Boosting the algorithm takes help of multiple weak algolrithms, changes their weights and builds on them, by which it becomes an powerful algorithm,this also adds up the results of all the weak algorithms and picks the best ones out of them and tries to reduce the loss function, due to which the acurracy would be boosted.

In [34]:
#importing Gradient Boost
from sklearn.ensemble import GradientBoostingClassifier

xgb_clf = GradientBoostingClassifier()
xgb_clf.fit(X, y)

GradientBoostingClassifier(criterion='friedman_mse', init=None,
              learning_rate=0.1, loss='deviance', max_depth=3,
              max_features=None, max_leaf_nodes=None,
              min_impurity_decrease=0.0, min_impurity_split=None,
              min_samples_leaf=1, min_samples_split=2,
              min_weight_fraction_leaf=0.0, n_estimators=100,
              n_iter_no_change=None, presort='auto', random_state=None,
              subsample=1.0, tol=0.0001, validation_fraction=0.1,
              verbose=0, warm_start=False)

In [36]:
#prediction based on validation data
y_valid1_pred=xgb_clf.predict(Xvalid)

In [35]:
#prediction based on test data
y_pred = xgb_clf.predict(X_test)

In [37]:
#Accuracy of Gradient Boost
print("Accuracy:",metrics.accuracy_score(yvalid, y_valid1_pred))

Accuracy: 0.8977


In [38]:
#confusiuon matrix of Gradient Boost
c_m =confusion_matrix(yvalid, y_valid1_pred) 
c_m

array([[35908,     1],
       [ 4091,     0]])

## Conclusion

We succesfully implemented Naive Bayes and Gradient Boost algorithms on the Santander data. By comparing the two models we can clearly see that Naive Bayes is better than gradient boost.Though the accuracies of both the models are close, we got an accuracy of 90.03% for Naive Bayes and 89.77% for Gradient Boost. Its better to use Naive Bayes in our case to predict the outcomes.

## References

3.2.4.3.5. sklearn.ensemble.GradientBoostingClassifier¶. (n.d.). Retrieved from https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingClassifier.html.